File: rgctl-tests/ecommerce-c/src/coolstore/services/promo_service.c
#include "ecommerce/coolstore/services/promo_service.h"
#include <string.h>

static double percent_off_for_item(const char *itemId) {
    if (itemId && strcmp(itemId, "329299") == 0) return 0.25;
    return -1.0;
}

void promo_apply_cart_item_promotions(shopping_cart_t *shoppingCart) {
    int i;
    if (!shoppingCart || shoppingCart->item_count == 0) return;
    for (i = 0; i < shoppingCart->item_count; i++) {
        shopping_cart_item_t *sci = &shoppingCart->shoppingCartItemList[i];
        double pct;
        if (!sci->has_product) continue;
        pct = percent_off_for_item(sci->product.itemId);
        if (pct >= 0) {
            sci->promoSavings = sci->product.price * pct * -1;
            sci->price = sci->product.price * (1 - pct);
        }
    }
}

void promo_apply_shipping_promotions(shopping_cart_t *shoppingCart) {
    if (!shoppingCart) return;
    if (shoppingCart->cartItemTotal >= 75) {
        shoppingCart->shippingPromoSavings = shoppingCart->shippingTotal * -1;
        shoppingCart->shippingTotal = 0;
    }
}

File: rgctl-tests/ecommerce-rust/src/coolstore/service/promo_service.rs
use std::collections::HashMap;

use crate::coolstore::model::ShoppingCart;

pub struct PromoService {
    percent_off_by_item: HashMap<String, f64>,
}

impl PromoService {
    pub fn new() -> Self {
        let mut percent_off_by_item = HashMap::new();
        percent_off_by_item.insert("329299".to_string(), 0.25);
        Self { percent_off_by_item }
    }

    pub fn apply_cart_item_promotions(&self, shopping_cart: &mut ShoppingCart) {
        if shopping_cart.shopping_cart_item_list.is_empty() {
            return;
        }
        for sci in &mut shopping_cart.shopping_cart_item_list {
            let Some(ref product) = sci.product else {
                continue;
            };
            if let Some(&pct) = self.percent_off_by_item.get(&product.item_id) {
                sci.promo_savings = product.price * pct * -1.0;
                sci.price = product.price * (1.0 - pct);
            }
        }
    }

    pub fn apply_shipping_promotions(&self, shopping_cart: &mut ShoppingCart) {
        if shopping_cart.cart_item_total >= 75.0 {
            shopping_cart.shipping_promo_savings = shopping_cart.shipping_total * -1.0;
            shopping_cart.shipping_total = 0.0;
        }
    }
}

impl Default for PromoService {
    fn default() -> Self {
        Self::new()
    }
}
